File: performance/performance_test.R
library(devtools)
library(igraph)
load_all("../R/")



# functions
#######################################################################

performance_trial = function(ns, k, motifs, method, nreps, graph_type){

  results = data.frame(
    n = integer(),
    k = double(),
    motif = character(),
    method = character(),
    time = double()
  )

  for(n in sort(ns, decreasing = TRUE)){
    for(motif in motifs){
      for(rep in 1:nreps){

        # graph parameters
        block_sizes = c(n)
        connection_matrix = matrix(k / n)

        # sample graph
        if(graph_type == "erdos_renyi"){
          adj_mat = sample_dsbm(block_sizes, connection_matrix)
        }
        else if(graph_type == "barabasi_albert"){
          sample_graph = sample_pa(
            n, k, directed = FALSE,
            start.graph = make_empty_graph(n = k, directed = FALSE)
          )
          adj_mat = drop0(as_adjacency_matrix(sample_graph))
        }

        if(method == "dense"){
          adj_mat = matrix(adj_mat, nrow = n)
        }

        # time mam construction
        t0 = Sys.time()
        build_motif_adjacency_matrix(adj_mat, motif, "func", "mean", method)
        t1 = Sys.time()
        dt = t1 - t0

        # add results
        results_row = data.frame("n" = n, "k" = k, "motif" = motif, "method" = method, "time" = dt, "rep" = rep)
        results = rbind(results, results_row)

        cat("n = ",n, ", k = ",k, ", ", motif, ", method = ", method,
              ", rep = ", rep, ", graph type = ", graph_type, ", time = ", round(dt, 3), "\n", sep = "")
        }
      }
    }

  # save results
  csv_name = paste("results/r_k", k, "_", method, "_", graph_type, ".csv", sep = "")
  write.csv(results, csv_name, quote = FALSE, row.names = FALSE)
}



# script
#######################################################################

motifs = c('M1','M8','M11')
nreps = 2

ns= c(101, 200, 500, 1000)
performance_trial(ns, 100, motifs, "dense", nreps, "barabasi_albert")
performance_trial(ns, 10, motifs, "dense", nreps, "barabasi_albert")
performance_trial(ns, 100, motifs, "dense", nreps, "erdos_renyi")
performance_trial(ns, 10, motifs, "dense", nreps, "erdos_renyi")

ns= c(101, 200, 500, 1000, 2000, 5000)
performance_trial(ns, 100, motifs, "sparse", nreps, "barabasi_albert")
performance_trial(ns, 10, motifs, "sparse", nreps, "barabasi_albert")
performance_trial(ns, 100, motifs, "sparse", nreps, "erdos_renyi")
performance_trial(ns, 10, motifs, "sparse", nreps, "erdos_renyi")

File: performance/performance_test.jl
using Graphs
using MotifCluster

function performance_trial(ns::Vector{Int}, k::Int, motifs::Vector{String},
        nreps::Int, graph_type::String)

    results = Dict{String, Any}[]
    for n in sort(ns, rev=true)
        for motif in motifs
            for rep in 1:nreps

                # graph parameters
                block_sizes = [n]
                connection_matrix = reshape([k/n], (1, 1))

                # sample graph
                if graph_type == "erdos_renyi"
                    adj_mat = sample_dsbm(block_sizes, connection_matrix, nothing, "unweighted")
                elseif graph_type == "barabasi_albert"
                    sample_graph = barabasi_albert(n, k)
                    adj_mat = adjacency_matrix(sample_graph)
                end

                # time mam construction
                t0 = time()
                build_motif_adjacency_matrix(adj_mat, motif, "func", "mean")
                t1 = time()
                dt = t1 - t0

                # add results
                result = Dict("n" => n, "k" => k, "motif" => motif, "rep" => rep, "time" => dt)
                push!(results, result)
                println("n = ", n, ", k = ", k, ", ", motif, ", rep = ", rep, ", graph type = ",
                        graph_type, ", time = ", round(dt, digits=3))

            end
        end
    end

    # save results
    filename = "results/julia_k" * string(k) * "_" * graph_type * ".csv"
    open(filename, "w") do file
        write(file, "n,k,motif,time,rep\n")
        for result in results
            write(file, string(result["n"]) * ",")
            write(file, string(k) * ",")
            write(file, result["motif"] * ",")
            write(file, string(result["time"]) * ",")
            write(file, string(result["rep"]) * "\n")
        end
    end

    return nothing
end

motifs = ["M1","M8","M11"]
nreps = 10

ns= [101, 200, 500, 1000]
performance_trial(ns, 100, motifs, nreps, "barabasi_albert")
performance_trial(ns, 10, motifs, nreps, "barabasi_albert")
performance_trial(ns, 100, motifs, nreps, "erdos_renyi")
performance_trial(ns, 10, motifs, nreps, "erdos_renyi")

ns= [101, 200, 500, 1000]
performance_trial(ns, 100, motifs, nreps, "barabasi_albert")
performance_trial(ns, 10, motifs, nreps, "barabasi_albert")
performance_trial(ns, 100, motifs, nreps, "erdos_renyi")
performance_trial(ns, 10, motifs, nreps, "erdos_renyi")
